### ISSUE TITLE
HA 2024.6.1 - async_unload - ValueError: list.remove(x): x not in list

### ISSUE THREAD

**Issue by thlucas1** [thlucas1/homeassistantcomponent_soundtouchplus#20]
### Describe the issue

With the HA 2024.6.1 release, I started getting `ValueError: list.remove(x): x not in list` errors in the System log whenever I change component options via the UI.

Did something change with the HA 2024.6 release with options update listener processing?  I use the following code in various integrations that provide options configuration:

```
async def async_setup_entry(hass:HomeAssistant, entry:ConfigEntry) -> bool:
...
        # register an update listener to reload configuration entry when options are updated.
        # this will return an "unlisten" function, which will be added to the configuration
        # "on_unload" event handler to automatically unregister the update listener when
        # the configuration is unloaded.
        listenerRemovePtr = entry.add_update_listener(options_update_listener)
        entry.async_on_unload(listenerRemovePtr)
...
async def options_update_listener(hass:HomeAssistant, entry:ConfigEntry) -> None:
...
```

### Reproduction steps

1. Assuming SoundTouchPlus component is installed.
2. via UI, change configuration options for the component and click 'Save'.
3. The `ValueError: list.remove(x): x not in list` exception will appear in the System Log.

### Debug logs

```text
Logger: homeassistant.config_entries
Source: config_entries.py:820
First occurred: 10:54:41 PM (2 occurrences)
Last logged: 10:55:20 PM

Error unloading entry Bose-ST300 for soundtouchplus
Error unloading entry Bose-ST10-1 for soundtouchplus
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/config_entries.py", line 820, in async_unload
    await self._async_process_on_unload(hass)
  File "/usr/src/homeassistant/homeassistant/config_entries.py", line 980, in _async_process_on_unload
    if job := self._on_unload.pop()():
              ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/config_entries.py", line 948, in <lambda>
    return lambda: self.update_listeners.remove(listener)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ValueError: list.remove(x): x not in list
```

**Comment by thlucas1** [comment]
@bdraco
Did something change with the HA 2024.6 release with options update listener processing? I use the following code in various integrations that provide options configuration:
```
async def async_setup_entry(hass:HomeAssistant, entry:ConfigEntry) -> bool:
...
        # register an update listener to reload configuration entry when options are updated.
        # this will return an "unlisten" function, which will be added to the configuration
        # "on_unload" event handler to automatically unregister the update listener when
        # the configuration is unloaded.
        listenerRemovePtr = entry.add_update_listener(options_update_listener)
        entry.async_on_unload(listenerRemovePtr)
...
async def options_update_listener(hass:HomeAssistant, entry:ConfigEntry) -> None:
...
```
With the HA 2024.6.1 release, I started getting ValueError: list.remove(x): x not in list errors in the System log whenever I change component options via the UI.

**Comment by bdraco** [comment]
I'm not aware of anything changing there

It looks like the only place they are removed is https://github.com/home-assistant/core/blame/2024.6.1/homeassistant/config_entries.py#L948

Is it possible the config entry is getting unloaded twice, or the callable returned from `add_update_listener` is being wrapped in `entry.async_on_unload` more than once?

**Comment by thlucas1** [comment]
I don't think it's getting unloaded twice.  The code is [here](https://github.com/thlucas1/homeassistantcomponent_soundtouchplus/blob/master/custom_components/soundtouchplus/__init__.py) if you want to have a quick look.

I do know there are no exceptions generated with the 2024.5.4 release of the same integration version.  The exceptions started appearing with the 2024.6 release, though I cannot be certain if it was with the 2024.6.0 or 2024.6.1 version.

Per my trace logs, it is only unloading the integration 1 time:
![image](https://github.com/thlucas1/homeassistantcomponent_soundtouchplus/assets/38439576/1a01aa25-0fd2-446c-b605-76009d23cc13)

**Comment by thlucas1** [comment]
@bdraco
I have the following code in my integration `async_unload_entry` method that checks if any update_listeners are still present:
```
            # a quick check to make sure all update listeners were removed (see method doc notes above).
            if len(entry.update_listeners) > 0:
                _logsi.LogArray(SILevel.Warning, "'%s': Component configuration update_listener(s) did not get removed before configuration unload (%d items - should be 0)" % (entry.title, len(entry.update_listeners)), entry.update_listeners)
                entry.update_listeners.clear()
```

Shouldn't the `update_listeners` array be empty by this point, as any entries within it have already been called before the `async_unload_entry` method is called?  I have the following notes in my code that indicate this (not saying they are correct, but that's the behavior I noticed when I put them in over a year ago):
```
    Note that any options update listeners (added via "add_update_listener") do not need 
    to be removed, as they are already removed by the time this method is called.
    This is accomplished by the "entry.async_on_unload(listener)" call in async_setup_entry,
    which removes them from the configuration entry just before it is unloaded.
```
Here's a trace log excerpt that indicates the `update_listeners` array is not empty at this point:
```
async_unload_entry (__init__.py, line 985)
'Bose-ST10-1': Component async_unload_entry configuration entry
'Bose-ST10-1': Component async_unload_entry is unloading our device instance from the domain
async_will_remove_from_hass (media_player.py, line 2433)
'Bose-ST10-1': MediaPlayer is stopping websocket notifications
Stopping SoundTouch web socket notification thread
Closing SoundTouch web socket connection
SoundTouch web socket event listener OnClose event: 'WebSocketClose' - (1000) b'STATUS_NORMAL (RC=1000) assumed - goodbye'
async_will_remove_from_hass (media_player.py, line 2451)
'Bose-ST10-1':Component async_unload_entry is removing our device instance data from the domain
'Bose-ST10-1': Component async_unload_entry unloaded configuration entry instance data
'Bose-ST10-1': Component configuration update_listener(s) did not get removed before configuration unload (1 items - should be 0)   <- ****** should not occur.
'Bose-ST10-1': Component async_unload_entry completed
```

**Comment by thlucas1** [comment]
@bdraco
I think I have this issue fixed, though I am still unclear as to the root cause.  To fix the `ValueError: list.remove(x): x not in list` exceptions, I removed the following line of code in my __init__.py `async_unload_entry` section:
```
entry.update_listeners.clear()
```

I know I specifically put this in my code (when I added configuration options support) to account for an update listener being called AFTER the integration entry was unloaded, and it was raising exceptions.  It seems that processing has changed somehow with the HA 2024.6 release, as I do not get the `ValueError: list.remove(x): x not in list` exceptions in the HA 2024.5.4 release.

Anyhow ... I commented that line of code and it removed the `ValueError: list.remove(x): x not in list` exceptions.  I also do not see any update listener exceptions in my trace log either, which is good.  

Thanks for listening, and hope I didn't waste too much of your time with this.
